### ISSUE TITLE
Create an overview page that compares repositories regarding multiple benchmarks

### ISSUE THREAD

**Issue by parttimenerd** [IPDSnelting/velcom#264]
Compare the current state of different repositories on a specified metric regarding one (or more) benchmarks by using a bar graph (or multiple bar graphs). Normalizing to a specific repository should also be possible.
This would allow to easily compare different approaches on multiple benchmarks. An example is the implementation of the overview graph from http://speedcenter.informatik.kit.edu/compprakt/comparison/:

![image](https://user-images.githubusercontent.com/490655/119486906-0a734e80-bd59-11eb-8fcc-ea78af1c433d.png)

**Comment by sgraf812** [comment]
It would be helpful if those bar charts were normalised according to a baseline. In fact, the bar chart from the OP is pretty useless as is: Nobody really cares for build times (except for timeouts, but nvm).

**Comment by parttimenerd** [comment]
![image](https://user-images.githubusercontent.com/490655/119622066-19183f00-be07-11eb-99d4-bfb7c34ba1ff.png)

**Comment by I-Al-Istannen** [comment]
What commit should this pick for each repository? The "newest" based on the committer date and a topological sort?

Do you ever want to look at older data and compare a few repositories at specific commits? Or can you safely ignore any commit besides the "newest"?

If you want to compare different strategies, wouldn't you rather want to compare multiple commits (e.g. branch tips) of a *single* repository instead of comparing multiple different repositories?

In summary: I am not sure I really understood the use case you have in mind.

**Comment by parttimenerd** [comment]
For the compiler lab, comparing the newest commit of every repository is enough.
You're right, that for other ocassions it might be useful to just be able to select multiple commits from different repositories.

**Comment by I-Al-Istannen** [comment]
Ah, I see. I don't know exactly which benchmarks you are performing there, but I guess we can assume the units are the same for every selected dimension? This means "output size" and "build time" will never be selected at the same time and we can therefore prohibit you from choosing benchmarks with different dimensions without rendering your use case impossible.

Looking at the codespeed page, some benchmarks seem to have vastly different value ranges (Building your Rust or Haskell compiler probably absolutely demolishes your Y-Axis). This isn't as bad when looking at the normalized views, so I hope its fine if we don't try too hard to reconcile that somehow.

Thanks for the clarification, I think I have a clearer idea of the problem this tries to solve now.

**Comment by parttimenerd** [comment]
For a starter, all units could be assumed to be same. Later one could add the support for multiple units (albeit only with normalization), omitting the units on the y axis.

**Comment by sgraf812** [comment]
Ratios don't have a unit (well, the empty unit), so that's probably a non-problem. There might be metrics of which it doesn't make sense to compute ratios, but let's not worry about that until it's a problem...

It would be great if we could pick benchmarks the ratios of which we are interested in.

Comparing HEAD of the tracked branch should be the most important use case.

**Comment by I-Al-Istannen** [comment]
I am not quite sure I am happy with the general layout, so I figured it wouldn't hurt to share. The page mostly steals components from other graphs, the repo selection can be found verbatim on the normal comparison page and the dimension selection on the repo detail page.

You can select multiple branches, but only a *single newest* datapoint will be returned and displayed - not one for every branch. This is consistent with the behaviour of the repo comparison graph.  
We also chose to display the newest commit and not the newest commit *we have a measurement for*. This sounded a bit more intuitive as you won't get stale data, but you will only get a useful graph if you've actually benchmarked that commit.

### Handled error conditions
Half the graph just tries to handle a plethora of annoying conditions. I am not quite sure about those, maybe you want to treat some differently? 
- Your baseline repo has no run. Hides graph under an error-overlay preventing interaction.
- Your baseline repo has no measurements for any of the selected dimensions. Hides graph under an error-overlay  preventing interaction.
- The newest commit in a selected repo might be unbenchmarked. Displays fake bars in every dimension with an "unbenchmarked" tooltip and label
- A repository might not have a measurement for a selected dimension. Displays a far bar with the "Not measured" label and tooltip
- The measurement for a selected dimension in a  repository might have failed. Displays a "Dimension failed" tooltip and shows the first N chars of the error message in the tooltip
- A run for a repo that is not selected as a baseline might have failed completely. Displays a fake bar in every dimension with a "Failed" label and the first N chars of the error message in the tooltip
- *And a **really ugly** case*: The baseline repo might not have a measurement for a dimension (either not measured or failed). In this case *we can not normalize* the other repositories and fall back to displaying their *absolute* values for the affected dimensions, likely absolutely thrashing the Y-Axis. 

### Alternative error handling solution
For all of the above we could instead omit the bar for the repository. This can be accomplished by giving it the equivalent of a 0 value, i.e. reserving space but not displaying anything, or just omitting the bar entirely.  
This will either leave a gap or absolutely no indication that and what is amiss, if we just silently drop the whole series entry for that repository.

If I were to beat Echarts a bit more than I currently do it would likely be possible to silently drop the column but still include it in the *tooltip*. This would keep the graph a bit tidier while still allowing you to figure what is amiss - but it makes it a lot harder to *actually notice* that a bar is missing somewhere.  
Nevertheless, this might be the most helpful display option if you expect to have many unbenchmarked, failed or unmeasured bars as the individual bars can get quite thin with a large amount of data.

## Further weird features
Echarts has quite a few random toolbox features, like zooming in on certain dimensions, adjusting the Y-Axis zoom using the slider on the right, showing data in a TSV-chart or adding automatically generated "decals" to the bars. We likely need to sacrifice the space up there anyways, so I decided to just throw them in there and see whether they are useful... That might not be what you should do in software development, but the cost for them was quite low at least.

### Demo
https://user-images.githubusercontent.com/20284688/120686335-2a191e00-c4a1-11eb-9eed-1154dad6a000.mp4

**Comment by parttimenerd** [comment]
Your demo looks quite good.

> We also chose to display the newest commit and not the newest commit we have a measurement for. This sounded a bit more intuitive as you won't get stale data, but you will only get a useful graph if you've actually benchmarked that commit.

I'm not quite sure whether this is a useful behavior: Repositories might be regularly updated, so displaying the last benchmarked commit might be more useful (albeit less intuitive), maybe adding a warning somewhere that the commit of Repo X is not the newest, might be better.

**Comment by Garmelon** [comment]
The queue prioritizes newer commits over older commits, so it shouldn't take long after a push until the newest commit is benchmarked again.

**Comment by parttimenerd** [comment]
I see your point, maybe @sgraf812 or @fread have an opinion on it too.

**Comment by sgraf812** [comment]
Wow, that looks great!

> You can select multiple branches, but only a single newest datapoint will be returned and displayed - not one for every branch. This is consistent with the behaviour of the repo comparison graph.

Hmm. The most recent commit in the repo (on some develop branch, say) is not necessarily the one that the team submits at the end of the lab. I suppose we simply shouldn't select the develop branch in addition to the main branch then. But during the lab, it's pretty useful to track multiple specific branches of the same repo, even to compare them to another. Does that make sense? Maybe you have an alternative workflow that works just as well?

Sorry to be a bit late to come forward with such a spec change.

Wrt. to error handling: sounds fine, I like what I see! Maybe you could pick a different baseline in the "ugly case", then it wouldn't be *that* ugly.

Oh, and as a stretch goal you might consider adding the option to switch to logarithmic/dB scale. Then increases will have positive bars while decreases will have negative bars. It's most useful when the spread is large (which might happen when some team wrote an optimisation that makes an asymptotic difference, or another compiler is just asymptotically stupid).

**Comment by I-Al-Istannen** [comment]
> Hmm. The most recent commit in the repo (on some develop branch, say) is not necessarily the one that the team submits at the end of the lab. I suppose we simply shouldn't select the develop branch in addition to the main branch then. But during the lab, it's pretty useful to track multiple specific branches of the same repo, even to compare them to another. Does that make sense? Maybe you have an alternative workflow that works just as well?

Yes, you can just not select the dev branch at the end then. You can visually compare multiple commits of a repo if they are close to each other in the mostly topological sorting VelCom uses on the detail page (e.g. [here](https://velcom.aaaaaaah.de/repo-detail/44bb5c8d-b20d-4bef-bdad-c92767dfa489?zoomXStart=1621996431396.0046&zoomXEnd=1622633005470.0789&dimensions=backend%3Acoverage%3Atodos%3Acheckstyle_violations%3A%3Afrontend%3Atodos&dayEquidistant=true)).
This view is helpful to look at multiple metrics over time, so it *might* solve that use case. It was mostly created to look at the evolution of your repository over time and shows the git DAG structure if you are zoomed in far enough. If you are actively developing some code, that might be enough? If it isn't, maybe augmenting [the comparison view](https://velcom.aaaaaaah.de/compare/44bb5c8d-b20d-4bef-bdad-c92767dfa489/to/44bb5c8d-b20d-4bef-bdad-c92767dfa489?hash1=8f8e506ddb477b14ef8d40332ee78f84ce2b5ff0&hash2=9069bf5e88bbff2b227b168e027e90d56e53fcd3) to show a barchart would solve that particular problem better? That would allow you to compare two arbitrary commits in a bar-charty way.

----

We could also show one commit per selected branch in the comparison view as you suggested. This would be inconsistent with [the other repo comparison](https://velcom.aaaaaaah.de/repo-comparison?zoomXStart=1622152800000&zoomXEnd=1623448800000&repos=44bb5c8d-b20d-4bef-bdad-c92767dfa489%3Amain%3A%3A10f9e8ab-8cf9-48e9-885b-c25b9b4dc058%3Amain%3A%3A9c85a30e-cc36-483a-aee2-453ff11b2c0b%3Afake-velcom-metrics&dimension=backend%3Abuild_time&dayEquidistant=true) (one dimension plotted over time in multiple repos), but it'd probably be easy enough to understand.
Coloring those bars would also be interesting, as the color is currently based on the repository. Maybe giving them the same color and different decals would be fine, so you can still tell what repository a given bar belongs to.


> Maybe you could pick a different baseline in the "ugly case", then it wouldn't be that ugly.

That's an interesting idea and would effectively constrain the height. I am not sure whether it would be confusing if the baseline is different for different dimensions. I might give it a shot and see whether I can somehow make that clear enough.

----

> Oh, and as a stretch goal you might consider adding the option to switch to logarithmic/dB scale. Then increases will have positive bars while decreases will have negative bars. It's most useful when the spread is large (which might happen when some team wrote an optimisation that makes an asymptotic difference, or another compiler is just asymptotically stupid).

That sounds like an interesting feature as well. Echarts natively supports a logarithmic axis, so that would be quite easy. A dB scale would be `10 log₁₀ (Value / Baseline)`?

**Comment by sgraf812** [comment]
> You can visually compare multiple commits of a repo if they are close to each other in the mostly topological sorting VelCom uses on the detail page (e.g. here).

Yes, that seems like the better workflow for that use case.

> I might give it a shot and see whether I can somehow make that clear enough.

That would be good, but I get your point about it might being confusing.

> A dB scale would be 10 log₁₀ (Value / Baseline)?

Something like that, yes.  Although using dB in such a scenario is a bit ... novel.

**Comment by I-Al-Istannen** [comment]
Hm. Echarts centers the log scale at 1, making this mostly a dB scale? It's a bit unfortunate, that the `1` bars disappear as their height becomes zero. Maybe I should fudge them by a bit.

![grafik](https://user-images.githubusercontent.com/20284688/121775254-d1bfdb80-cb86-11eb-843c-62a9c1086671.png)

**Comment by I-Al-Istannen** [comment]
This wasn't accidentally closed, but it would probably be a good idea if you played around a bit with it on the https://velcom.aaaaaaah.de/ instance (a few hours after this comment preferably :P).

If you find anything noteworthy (maybe on your own toy instance), feel free to comment here and close the issue after you are happy :)

**Comment by I-Al-Istannen** [comment]
@parttimenerd do you have any plan for testing this or are you going to wait until fall?

**Comment by parttimenerd** [comment]
Looks nice. Just three remarks:
1. what does "Show decals" do?
2. please make it usable without selecting a base case
3. I see values for repos with bars that tell me, that a run failed, why do I see any data there?

**Comment by I-Al-Istannen** [comment]
> what does "Show decals" do?

"Show decals" shows decals, a term Echarts borrowed from computer graphics apparently. In this case it applies a pattern to the bars:
![grafik](https://user-images.githubusercontent.com/20284688/124601809-8e484c80-de68-11eb-95ad-835b1b25f02c.png)
This can help if the colors are similar (or you are color blind, your monitor isn't great, etc.) and as Echarts provided that feature out of the box I exposed it in the UI.

----
> It seems not to be usable without selecting a base case

Correct. This is by design after sgraf's comments (e.g. [here](https://github.com/IPDSnelting/velcom/issues/264#issuecomment-848888048)). If you do not normalize the display, the different dimensions might make comparing and sharing a Y-Axis useless - their value range can vary considerably. I thought the consensus was to force normalization and the first prototype I shared in the video required that as well.

I could lift that requirement, the chart component can also handle absolute values.

----
> I see values for repos with bars that tell me, that a run failed, why do I see any data there?

We needed to find some compromise for those cases. Not displaying *anything* would be confusing to users ("I selected this dimension and the commit is benchmarked, why is there no bar here?"). I explored other options in [this comment](https://github.com/IPDSnelting/velcom/issues/264#issuecomment-854070646) in the "Alternative error handling solution" section.
If you believe one of those would be more intuitive, I could adjust it.
It is quite easy to tell in the current solution that something's wrong and my assumption was that measurements should not fail normally, making this an exceptional event you might want to explicitly see.

**Comment by parttimenerd** [comment]
> "Show decals" shows decals, a term Echarts borrowed from computer graphics apparently. In this case it applies a pattern to the bars:
Interesting, maybe make the title text a bit more expressive (e.g. "Add patterns to bars").

> I could lift that requirement, the chart component can also handle absolute values.
I would appreciate it. This is helpful to get a glimpse on the actual values.

> We needed to find some compromise for those cases. Not displaying anything would be confusing to users ("I selected this dimension and the commit is benchmarked, why is there no bar here?"). I explored other options in this comment in the "Alternative error handling solution" section.
I see your point. It just gives the last value it measured?

**Comment by I-Al-Istannen** [comment]
>  Interesting, maybe make the title text a bit more expressive (e.g. "Add patterns to bars").

Yea, that's probably a good idea then. I went with the terminology echarts uses but that is probably not really a common term :)

>  I would appreciate it. This is helpful to get a glimpse on the actual values.

Alright, I will allow it then and make it the user's responsibility to only select comparable dimensions.

>  I see your point. It just gives the last value it measured?

As the endpoint only returns the most recent benchmark it doesn't know that, instead it will pick a suitable high value (I think the max of the values currently) so it is visible. I'm open for suggestions though, I am not perfectly happy with it but the other options aren't really much better if we assume errors are relevant.

**Comment by parttimenerd** [comment]
> As the endpoint only returns the most recent benchmark it doesn't know that, instead it will pick a suitable high value (I think the max of the values currently) so it is visible. I'm open for suggestions though, I am not perfectly happy with it but the other options aren't really much better if we assume errors are relevant.
Yes, this is a problem. I don't know any good solution, so maybe we just wait till people really use it (e.g. till the compiler lab starts in fall).

**Comment by I-Al-Istannen** [comment]
Is #283 what you had in mind? I chose to keep highlighting it in red because you likely *want* to select something there and we can no longer provide a visual clue in the overlay.

**Comment by parttimenerd** [comment]
Looks good.

**Comment by sgraf812** [comment]
The few times so far I had a use case for the feature, I was very pleased. Great work!

In https://github.com/IPDSnelting/velcom/issues/264#issuecomment-862533008 you said we should try the instance at https://velcom.aaaaaaah.de/ and I was hoping to find the work on log scale https://github.com/IPDSnelting/velcom/issues/264#issuecomment-860054787, but couldn't find a button to press. Fair enough, maybe we should move it into a separate issue anyway.

I think we can close this issue now. I'm quite happy with what we got.

**Comment by I-Al-Istannen** [comment]
Sounds good :)

> Fair enough, maybe we should move it into a separate issue anyway

I wasn't really sold on the way it looked and how best to handle a reference (as 1 becomes 0 and disappears in a log scale). If you have an opinion about that and think a log is nice, feel free to open a new issue :)

**Comment by sgraf812** [comment]
> If you have an opinion about that and think a log is nice, feel free to open a new issue :)

Will do, if/when that is the case.
